server/src/main/java/io/unitycatalog/server/service/VolumeService.java

```diff
@@ -2,53 +2,175 @@
+    initializeAuthorizations(volumeInfo);
+    return HttpResponse.ofJson(volumeInfo);
   }
 
   @Get("")
+  @AuthorizeExpression("#defer")
   public HttpResponse listVolumes(
       @Param("catalog_name") String catalogName,
       @Param("schema_name") String schemaName,
       @Param("max_results") Optional<Integer> maxResults,
       @Param("page_token") Optional<String> pageToken,
       @Param("include_browse") Optional<Boolean> includeBrowse) {
-    return HttpResponse.ofJson(
-        VOLUME_REPOSITORY.listVolumes(
-            catalogName, schemaName, maxResults, pageToken, includeBrowse));
+    ListVolumesResponseContent listVolumesResponse = VOLUME_REPOSITORY.listVolumes(
+            catalogName, schemaName, maxResults, pageToken, includeBrowse);
+
+    filterVolumes("""
+            #authorize(#principal, #metastore, OWNER) ||
+            (#authorizeAll(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG)) ||
+            (#authorize(#principal, #schema, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #volume, OWNER, READ_VOLUME))
+            """, listVolumesResponse.getVolumes());
+
+    return HttpResponse.ofJson(listVolumesResponse);
   }
 
   @Get("/{full_name}")
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          (#authorizeAny(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #volume, OWNER, READ_VOLUME))
+          """)
+  @AuthorizeKey(METASTORE)
   public HttpResponse getVolume(
-      @Param("full_name") String fullName,
+      @Param("full_name") @AuthorizeKey(VOLUME) String fullName,
       @Param("include_browse") Optional<Boolean> includeBrowse) {
     return HttpResponse.ofJson(VOLUME_REPOSITORY.getVolume(fullName));
   }
 
   @Patch("/{full_name}")
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          (#authorize(#principal, #volume, OWNER) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA))
+          """)
+  @AuthorizeKey(METASTORE)
   public HttpResponse updateVolume(
-      @Param("full_name") String fullName, UpdateVolumeRequestContent updateVolumeRequest) {
+      @Param("full_name") @AuthorizeKey(VOLUME) String fullName, UpdateVolumeRequestContent updateVolumeRequest) {
     return HttpResponse.ofJson(VOLUME_REPOSITORY.updateVolume(fullName, updateVolumeRequest));
   }
 
   @Delete("/{full_name}")
-  public HttpResponse deleteVolume(@Param("full_name") String fullName) {
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          (#authorize(#principal, #volume, OWNER) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA))
+          """)
```

Comment:
Catalog owner, schema owner with USE_CATALOG, and volume owner with USE_CATALOG and USE_SCHEMA can delete a volume.